### PR TITLE
build-support/php: reorganize files

### PR DESCRIPTION
**Pull request by drupol**
Pave the way for having multiple versions in the future

The changes in this PR doesn't alter any packages at all, it just move files around.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by drupol** [comment]
@ofborg build php.packages.composer phpunit

File: pkgs/build-support/php/builders/default.nix [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{ callPackage, ... }:
+{
+  v1 = {
+    buildComposerProject = callPackage ./v1/build-composer-project.nix { };
+    mkComposerRepository = callPackage ./v1/build-composer-repository.nix { };
+    composerHooks = callPackage ./v1/hooks { };
```
Comment (jtojnar) [inline pkgs/build-support/php/builders/default.nix:6]: `callPackage` should import a file that contains a function that returns a single derivation.

File: pkgs/top-level/php-packages.nix [hunk 2]
```diff
@@ -46,15 +46,15 @@
 , fetchpatch
 }:
 
+let
+  builders = callPackages ../build-support/php/builders { };
```
Comment (jtojnar) [inline pkgs/top-level/php-packages.nix:50]: `callPackage`, on the other hand, should import a file that contains a function that returns a single derivation.

File: pkgs/build-support/php/builders/default.nix (hunk 1 above)
Comment (drupol) [reply pkgs/build-support/php/builders/default.nix]: Thanks, I just fixed this one.

File: pkgs/top-level/php-packages.nix (hunk 2 above)
Comment (drupol) [reply pkgs/top-level/php-packages.nix:50]: Should I use `callPackage` here ?

Comment (jtojnar) [reply pkgs/top-level/php-packages.nix:50]: I would use plain `import`.

Comment (jtojnar) [reply pkgs/top-level/php-packages.nix:50]: Indeed. In addition to auto-wiring, the point of `callPackage{s,}` is to set up `.override`. That will not really work here anyway due to the nested structure.

**Comment by drupol** [comment]
@jtojnar Can you help with the current eval issue?

**Comment by drupol** [comment]
@ofborg build php php.packages.psalm phpunit php.extensions.dom

File: pkgs/top-level/php-packages.nix [hunk 3]
```diff
@@ -1,5 +1,6 @@
 { stdenv
 , config
+, callPackage
```
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/top-level/php-packages.nix]: I guess it's fixed now.

File: pkgs/top-level/php-packages.nix [hunk 4]
```diff
@@ -46,15 +47,16 @@
 , fetchpatch
 }:
 
-lib.makeScope pkgs.newScope (self: with self; {
+lib.makeScope pkgs.newScope (self: with self; let
```
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/top-level/php-packages.nix:49]: That's fixed too, I think.

File: pkgs/top-level/php-packages.nix [hunk 5]
```diff
@@ -46,15 +46,19 @@
...
+  builders = import ../build-support/php/builders {
+    inherit callPackages buildPecl;
+    inherit (self) callPackage;
+  };
+in {
+  inherit buildPecl;
 
-  mkComposerRepository = callPackage ../build-support/php/build-composer-repository.nix { };
-  buildComposerProject = callPackage ../build-support/php/build-composer-project.nix { };
+  inherit (builders.v1) buildComposerProject composerHooks mkComposerRepository;
```
Comment (jtojnar) [inline pkgs/top-level/php-packages.nix:61]: Why not just:

```suggestion
lib.makeScope pkgs.newScope (self: with self; let
  builders = import ../build-support/php/builders {
    inherit callPackages callPackage buildPecl;
  };
in {
  buildPecl = callPackage ../build-support/php/build-pecl.nix {
    php = php.unwrapped;
  };

  inherit (builders.v1) buildComposerProject composerHooks mkComposerRepository;
```

And possibly replacing the `with self;` with `let inherit (self) /* … */;` in a preceding commit.

Comment (drupol) [reply pkgs/top-level/php-packages.nix]: That's the part where I'm confused and where I lack of XP.

> And possibly replacing the with self; with let inherit (self) /* … */; in a preceding commit.

What am I inheriting here? What are the attributes?

Comment (jtojnar) [reply pkgs/top-level/php-packages.nix]: I would start with php and add more stuff until the evaluation errors disappear.

**Comment by drupol** [comment]
@ofborg build php php.packages.psalm phpunit php.extensions.dom

**Review by jtojnar — APPROVED** [review]

**Pull request merged**